server/service/src/programs/patient/patient_updated.rs

```diff
@@ -11,12 +11,41 @@ use super::{
     UpdatePatientError,
 };
 
-/// Callback called when the document has been updated
-pub fn patient_document_updated(
+// create name_store_join if not existing
+pub fn create_patient_name_store_join(
     con: &StorageConnection,
     store_id: &str,
+    name_id: &str,
+) -> Result<(), UpdatePatientError> {
+    let name_repo = NameRepository::new(con);
+    let name = name_repo.query_one(
+        store_id,
+        NameFilter::new()
+            .is_customer(true)
+            .id(EqualFilter::equal_to(name_id)),
+    )?;
+    if name.is_none() {
+        // add name store join
+        let name_store_join_repo = NameStoreJoinRepository::new(con);
+        name_store_join_repo.upsert_one(&NameStoreJoinRow {
+            id: uuid(),
+            name_id: name_id.to_string(),
+            store_id: store_id.to_string(),
+            name_is_customer: true,
+            name_is_supplier: false,
+            is_sync_update: false,
+        })?;
+    }
+    Ok(())
+}
+
+/// Callback called when a patient document has been updated
+/// Updates the names table for the updated patient.
+pub fn update_patient_row(
```

Comment:
I know it's more lines of code, but I really like the use of destructuring and spread in cases like this, due to exhaustive match and visibility of unused fields

```diff
diff --git a/server/service/src/programs/patient/patient_updated.rs b/server/service/src/programs/patient/patient_updated.rs
index 189710f74..95b41c892 100644
--- a/server/service/src/programs/patient/patient_updated.rs
+++ b/server/service/src/programs/patient/patient_updated.rs
@@ -44,31 +44,45 @@ pub fn create_patient_name_store_join(
 pub fn update_patient_row(
     con: &StorageConnection,
     update_timestamp: &DateTime<Utc>,
-    patient: SchemaPatient,
+    SchemaPatient {
+        code,
+        code_2,
+        contact_details,
+        date_of_birth,
+        first_name,
+        gender,
+        id,
+        is_deceased,
+        last_name,
+        middle_name: _,
+        notes: _,
+        passport_number: _,
+        socio_economics: _,
+        allergies: _,
+        birth_place: _,
+        health_center: _,
+        date_of_birth_is_estimated: _,
+        date_of_death: _,
+        family: _,
+    }: SchemaPatient,
     is_sync_update: bool,
 ) -> Result<(), UpdatePatientError> {
-    let contact = patient.contact_details.get(0);
-    let date_of_birth = match patient.date_of_birth {
+    let contact = contact_details.get(0);
+    let date_of_birth = match date_of_birth {
         Some(date_of_birth) => Some(NaiveDate::from_str(&date_of_birth).map_err(|err| {
             UpdatePatientError::InternalError(format!("Invalid date of birth format: {}", err))
         })?),
         None => None,
     };
-    let address = patient.contact_details.get(0);
+    let address = contact_details.get(0);
     let name_repo = NameRowRepository::new(con);
-    let existing_name = name_repo.find_one_by_id(&patient.id)?;
-    let existing_name = existing_name.as_ref();
+    let existing_name = name_repo.find_one_by_id(&id)?;
+
     name_repo.upsert_one(&NameRow {
-        id: patient.id.clone(),
-        name: patient_name(&patient.first_name, &patient.last_name),
-        code: patient.code.unwrap_or("".to_string()),
         r#type: NameType::Patient,
-        is_customer: existing_name.map(|n| n.is_customer).unwrap_or(true),
-        is_supplier: existing_name.map(|n| n.is_supplier).unwrap_or(false),
-        supplying_store_id: existing_name.and_then(|n| n.supplying_store_id.clone()),
-        first_name: patient.first_name,
-        last_name: patient.last_name,
-        gender: patient.gender.and_then(|g| match g {
+        name: patient_name(&first_name, &last_name),
+        is_deceased: is_deceased.unwrap_or(false),
+        gender: gender.and_then(|g| match g {
             SchemaGender::Female => Some(Gender::Female),
             SchemaGender::Male => Some(Gender::Male),
             SchemaGender::Transgender => Some(Gender::Transgender),
@@ -77,24 +91,28 @@ pub fn update_patient_row(
             SchemaGender::Unknown => Some(Gender::Unknown),
             SchemaGender::NonBinary => Some(Gender::NonBinary),
         }),
-        date_of_birth,
-        charge_code: existing_name.and_then(|n| n.charge_code.clone()),
-        comment: existing_name.and_then(|n| n.comment.clone()),
         country: address.and_then(|a| a.country.clone()),
         address1: address.and_then(|a| a.address_1.clone()),
         address2: address.and_then(|a| a.address_2.clone()),
         phone: contact.and_then(|c| c.phone.clone().or(c.mobile.clone())),
         email: contact.and_then(|c| c.email.clone()),
         website: contact.and_then(|c| c.website.clone()),
-        is_manufacturer: existing_name.map(|n| n.is_manufacturer).unwrap_or(false),
-        is_donor: existing_name.map(|n| n.is_donor).unwrap_or(false),
-        on_hold: existing_name.map(|n| n.on_hold).unwrap_or(false),
-        created_datetime: existing_name
-            .and_then(|n| n.created_datetime.clone())
-            .or(Some(update_timestamp.naive_utc())), // assume there is no earlier doc version
-        is_deceased: patient.is_deceased.unwrap_or(false),
-        national_health_number: patient.code_2,
+
+        id,
+        first_name,
+        last_name,
+        date_of_birth,
+        national_health_number: code_2,
         is_sync_update,
+        code: code.unwrap_or_default(),
+
+        ..existing_name.unwrap_or(NameRow {
+            is_customer: true,
+            is_deceased: false,
+            is_donor: true,
+            created_datetime: Some(update_timestamp.naive_utc()), // assume there is no earlier doc version
+            ..NameRow::default()
+        })
     })?;
 
     Ok(())
```

Above would suggest that the following fields are considered `common` omSupply fields that are not in mSupply, would they be common for all countries ? (or are we still to add patient schema to configurable schema and serialise this document to a NameRow type structure ?)

I still quite strong on keeping the know at runtime schema strongly typed (i.e. for both graphql and sync)

Reply:
Thanks yes agree with the destructuring. I updated the code.

The SchemaPatient is the common base schema which can still be adapted for different countries. E.g. this schema + country specific values such as birth order or handedness. Here, we receive the full patient document but we only extract the base schema for the integration.

Is there an option to have additional fields in GraphQL, e.g. a JSON input object with some predefined fields. This would be pretty neat for the API. You can do this in REST/swagger but haven't found something similar in GraphQL?